### ISSUE TITLE
feat(model): implement deferred prefill quantization for KV cache to eliminate prompt rotation overhead (rotorquant borrow)

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12274]
```routing
lane: model
paths: ["internal/model/**", "internal/ctxmmu/**"]
expected_steps: 6
```

# feat(model): implement deferred prefill quantization for KV cache to eliminate prompt rotation overhead (rotorquant borrow)

## Problem

For LLM prefill passes over long system prompts, rotating and quantizing every KV vector token-by-token during the forward pass degrades Time-to-First-Token (TTFT) by up to 5x and compounds floating-point roundoff errors across thousands of context tokens. Today prefill throughput collapses from $>6,000$ tok/s to $<800$ tok/s when online KV quantization is enabled. Better because RotorQuant's deferred quantization stages prompt KV vectors in uncompressed FP16/BF16 throughout prefill, executes a single bulk GPU/SIMD quantization conversion post-prefill, and switches to single-token quantization only during decode. The witness benchmarks prefill TTFT with deferred quantization and proves prefill latency matches native FP16 within 3% while VRAM footprint drops to compressed size before decode starts.

## Centrality and P1-P4

Centrality: Core
P1: advanced - restores prefill speed to native FP16 performance.
P2: advanced - eliminates quantization error compounding across long system prompts.
P3: advanced - clear separation between prompt staging and autoregressive decode pools.
P4: advanced - preserves high prompt-cache hit rates without numerical drift.

## Scope

Add a deferred prefill staging buffer to `internal/model` and `internal/ctxmmu`, triggering bulk page quantization upon prompt processing completion.

## Gold-plating boundary

No multi-GPU tensor-parallel redistribution during conversion; single-device forward pass only.

## Done condition

- [ ] `internal/model` stages prompt KV tokens in FP16/BF16 memory during multi-token prefill ($N > 1$).
- [ ] Post-prefill conversion kernel converts staging blocks to target KV quant dtype in a single batched pass.
- [ ] Autoregressive decode ($N = 1$) writes directly to quantized blocks.
- [ ] Tests verify output token identity between deferred quantization and non-deferred baseline.

## Verifiable Witness

```text
go test -v ./internal/model -run TestDeferredPrefillQuantization -count=1
```

## Prior-art source and disposition

- Target: `scrya-com/rotorquant` @ `fcd7676` (`cpy-planar-iso.cu`, `llama-cpp-turboquant`).
- License: MIT. Disposition: INSPIRE-ONLY (clean-room Go implementation).

## Likely files

- `internal/model/prefill_quant.go`
- `internal/model/prefill_quant_test.go`
- `internal/ctxmmu/kv_pool.go`

**Comment by anthony-chaudhary** [comment]
Premise audit at public trunk `430fd2b2c71fd23c6600c41f0f4b32eaceb56401` (2026-09-12): deferred KV quantization is not yet a supported performance candidate on the inspected native host/Vulkan paths.

- `internal/model/kvcache.go:11` stores K, Kraw, and V as FP32 slices. `internal/model/kv.go:319` explicitly says Q8 weight selection keeps the same FP32 KV cache. `internal/model/quant_forward.go` appends FP32 rows.
- `internal/compute/vulkan_qwen35_sequence.go:593-605` computes FP32 RoPE, appends raw keys/rotated keys/values, then immediately consumes the rotated keys in FP32 causal attention. `internal/compute/vulkan_kv_vulkan.go:80` allocates FP32 cache storage and `growAppend` copies four bytes per element. This required positional RoPE is distinct from a quantizer's rotation and cannot simply move after prefill.
- `internal/ctxmmu/kv_pool.go:419` copies already encoded bytes into pages; it does not perform online rotation or quantization. The standalone `NewAsymmetricKVCache`/`QuantizeAsymmetricKV` implementation has no production callers outside its own file in the inspected internal tree.

Source links: [model cache](https://github.com/anthony-chaudhary/fak/blob/430fd2b2c71fd23c6600c41f0f4b32eaceb56401/internal/model/kvcache.go#L11), [weight/KV distinction](https://github.com/anthony-chaudhary/fak/blob/430fd2b2c71fd23c6600c41f0f4b32eaceb56401/internal/model/kv.go#L319), [Vulkan sequence](https://github.com/anthony-chaudhary/fak/blob/430fd2b2c71fd23c6600c41f0f4b32eaceb56401/internal/compute/vulkan_qwen35_sequence.go#L593), [page write](https://github.com/anthony-chaudhary/fak/blob/430fd2b2c71fd23c6600c41f0f4b32eaceb56401/internal/ctxmmu/kv_pool.go#L419).

The issue's 5x and 6000-to-800 tok/s figures are not fak measurements established by this audit. No staging buffer or conversion candidate was introduced, and no additional context memory was allocated by a candidate. Conversion overhead and net full-prefill uplift remain unmeasured, not zero.

Required next slice: identify and witness an eligible native compressed-KV read/write path before comparing online with deferred conversion. Its admission must account for existing resident state, bounded staging, destination and conversion scratch, including allocation overlap and checked arithmetic; refuse insufficient memory before allocation and stop on OOM. Compare identical source/model/backend/prompt/quality envelopes; verify final compressed state and generated tokens; separately time conversion and repeat complete prefill measurements. Keep a candidate only when those repeated complete measurements show net benefit. Physical criteria remain open; software tests cannot qualify them.

This was a read-only investigation in the assigned Codex worktree; no owned source paths, code commit, lease takeover, or shared shim changes. A matching local ticket was not found through the available ticket stores/docs searches. Keep #12274 open as the durable implementation/physical-witness tracker.

Executed software witness (WSL, Go 1.26.6, assigned checkout, no mirror): `FAK_FAST=0 ./test.ps1 ./internal/model -v -run 'TestCachedDecodeMatchesPrefill|TestQwen35HybridQuantBatchedPrefillMatchesTokenLoop' -count=1 -timeout=90s`.

- PASS: `TestQwen35HybridQuantBatchedPrefillMatchesTokenLoop` (1.31s); selected package run PASS (1.457s). This exercises synthetic native Q8-weight prefill versus the token loop and checks logits/cache state; it does not test deferred compression or establish physical performance.
- SKIP: `TestCachedDecodeMatchesPrefill`, because the exported `.cache/smollm2-135m` weights are absent. No weight-backed output-identity claim is made.

**Comment by anthony-chaudhary** [comment]
Landed bounded software leaf in fede30b3ba (fak main): deferred prefill quantization bulk pass for the Q8 KV cache. appendBatchedKV now routes the whole P-row prefill panel through kvPackedRow.appendRowsBulk (one QuantizeKVQ8_0 call) instead of P per-row passes with P Scale+Codes allocations. Witness: go test ./internal/model -run TestDeferredPrefillQuantization -count=1 -> ok; BenchmarkDeferredPrefillQuantizationBulk -> bulk 675us/4 allocs vs per-row 855us/1053 allocs per 512x128 panel (2026-09-20, [SW-VERIFIED]). Boundary 5/5 PASS; Separation-Verdict SINGLE_PROVIDER.

Reopening per the 2026-09-12 premise audit: this issue remains the durable tracker for the FULL goal, which this leaf does not satisfy. Still open: (a) witnessing an eligible native compressed-KV read/write path where deferred conversion is a net win; (b) accounting for staging + conversion scratch under bounded memory with refuse-before-allocate; (c) matched source/model/backend/prompt/quality-envelope prefill measurements proving net benefit; (d) the >=3% prefill-TTFT-vs-FP16 and VRAM-before-decode criteria on physical gfx1151 hardware. No fabricated throughput claim is made from the in-memory benchmark.

**Comment by anthony-chaudhary** [comment]
Resolved on trunk by commit fede30b3ba (perf(model): defer prefill KV quantization to one bulk pass (fak model)), which carries Fixes #12274.

Verification witness (re-run against current origin/main at e3bd1a4c79):
- `go test ./internal/model -run TestDeferredPrefillQuantization -count=1` -> ok (TestDeferredPrefillQuantizationBulkMatchesPerRow, TestDeferredPrefillQuantizationMultiRowWidths, TestDeferredPrefillQuantizationPrefillIsStable all PASS)
- `go vet ./internal/model` -> clean (exit 0)

Deliverable is real on the prefill hot path: `kvPackedRow.appendRowsBulk` (internal/model/kvcache_q8.go:64) is called from `KVCache.appendBatchedKV` at kvcache_q8.go:460-461, converting the whole P-row prefill panel in one QuantizeKVQ8_0 pass instead of P per-row passes. Qualification: [SW-VERIFIED] only; no physical KEEP/REJECT is claimed.

Re-witnessed by the scheduled Halo-perf issue-orchestrator pass (single provider slot, 2026-09-22).
